### DISCUSSION TITLE
Availability for publications, links and licenses in ODL

### DISCUSSION THREAD

**Discussion by HadrienGardeur** [opds-community/drafts#63] [Ideas]
In ODL, a number of stakeholders have the requirement to convey the status of a publication or a specific copy acquired by a library.
Among those use-cases we could list:

- Open-access or public domain titles that are removed from distribution
- Books that get removed for legal reasons
- Copies bought by mistake by a library, for which they're issued a refund
- Copies tied to a publication with a problematic file, where a refund is issued instead of replacing the file

The concept of a "status" is not completely foreign to ODL, but it doesn't really cover these use cases:

- each loan has a status as defined in https://readium.org/lcp-specs/releases/lsd/latest.html#23-status-of-a-license
- each copy/license acquired by a library also has a status, which is strictly conveyed in the License Info Document: https://drafts.opds.io/odl-1.0#4-license-info-document

Among the question that should be tackled:

- Should we extend the value listed in https://drafts.opds.io/odl-1.0#4-license-info-document ?
- Is there a requirement for additional information in addition to a status? For example, with a refund, do we need a specific reason why the copy was refunded?
- How can we express this info in an ODL feed directly?
- Should we use the same field/values for the status of a publication? How should we convey/transmit this information for a publication?

cc @pbrantley @jce1028 @jonathangreen

**Comment by HadrienGardeur** [comment]
The current list of known-values for the status of a loan are:

- `ready`: The License Document is available, but the user hasn’t accessed the License and/or Status Document yet.
- `active`: The license is active, and a device has been successfully registered for this license.This is the default value if the License Document does not contain a registration link, or a registration mechanism through the license itself.
- `revoked`: The license is no longer active, it has been invalidated by the Issuer.
- `returned`: The license is no longer active, it has been invalidated by the User.
- `cancelled`: The license is no longer active because it was cancelled prior to activation.
- `expired`: The license is no longer active because it has expired.

The same list for the status of a copy contains:

- `preorder`
- `available`
- `unavailable`

None of them feel like a proper solution for this problem.

**Comment by jce1028** [comment]
Im not sure if it was captured in your articulation of the use cases but an
example of a title being no longer available in a collection is that it was
one of many titles in a subscribed-to collection that is replaced with
another title.  This scenario covers subscriptions to content that the
library gets from a subscription service but the library is not selecting
individual titles - the library is getting access to the titles as part of
a general subscription to a vendor collection.

On Wed, Mar 15, 2023 at 12:16 PM Hadrien Gardeur ***@***.***>
wrote:

> In ODL, a number of stakeholders have the requirement to convey the status
> of a publication or a specific copy acquired by a library.
> Among those use-cases we could list:
>
>    - Open-access or public domain titles that are removed from
>    distribution
>    - Books that get removed for legal reasons
>    - Copies bought by mistake by a library, for which they're issued a
>    refund
>    - Copies tied to a publication with a problematic file, where a refund
>    is issues instead of replacing the file
>
> The concept of a "status" is not completely foreign to ODL, but it doesn't
> really cover these use cases:
>
>    - each loan has a status as defined in
>    https://readium.org/lcp-specs/releases/lsd/latest.html#23-status-of-a-license
>    - each copy/license acquired by a library also has a status, which is
>    strictly conveyed in the License Info Document:
>    https://drafts.opds.io/odl-1.0#4-license-info-document
>
> Among the question that should be tackled:
>
>    - Should we extend the value listed in
>    https://drafts.opds.io/odl-1.0#4-license-info-document ?
>    - Is there a requirement for additional information in addition to a
>    status? For example, with a refund, do we need a specific reason why the
>    copy was refunded?
>    - How can we express this info in an ODL feed directly?
>    - Should we use the same field/values for the status of a publication?
>    How should we convey/transmit this information for a publication?
>
> cc @pbrantley <https://github.com/pbrantley> @jce1028
> <https://github.com/jce1028> @jonathangreen
> <https://github.com/jonathangreen>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/opds-community/drafts/discussions/63>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIE7W3NQWESI4C6DYY7YQDW4HTMPANCNFSM6AAAAAAV4CATTA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Regards,
James English

www.linkedin.com/in/jamesenglish

**Reply by jce1028** [reply]
So a specific Use case is at the publication level vs license level since we deal with publications that are part of a collection or provider integration
For example a - 
A Subscription Ebook service provider makes a collection of books available in a OPDS Catalog feed  that can be accessed by a library. The titles that constitute the collection may come and go based on some determination or decision by the provider - not the library.  The library has simply purchased a subscription to a whole collection which was defined by the service provider. For that reason a title may be removed at any time for the following - 

- The number of uses have been consumed
- The titles time of availability within the collection has passed
- The title is no longer available in the collection (for some reason like selection profitability by provider or other opaque reason)
 -The title is no longer allowed for distribution by provider due to some rights dispute or contract expiration with the right holder or publisher
 -The title was accidentally added by mistake and needs to be removed by the consuming feed client.

**Reply by jce1028** [reply]
Our use case is for harvesting a collection from another service so we can republish those items into a combined catalog of items from various distributors or collections provided to the client app.

**Comment by HadrienGardeur** [comment]
Going back to this discussion, I think we'll hit a wall if we try to cover every single reason why something gets removed through the use of status values.

Here's my preferred approach:

- clearly mark each publication, license or acquisition link as removed
- use a URI to identify the reason why something has been removed
- and also allow a human readable string

This could look like that:
```json
"removed": {
  "reason": "https://example.com/reason/refund",
  "detail": "This publication was refunded due to a faulty EPUB that doesn't meet accessibility criteria"
} 
```

We could define a list of well-known reasons through a registry but this would also allow catalog providers to define their own reasons using a URL.

We might also consider adding a specific time/date for this removal.

**Reply by HadrienGardeur** [reply]
Here are a few examples as well, to illustrate how this could work in practice.

**Example 1: the whole publication has been removed**

**Note:** Unlike OAI-PMH or the Atom tombstone draft, I wouldn't recommend dropping all the relevant info about a publication, which means that the requirement to have a valid OPDS publication would remain the same. That said, catalog providers could send minimal publications instead with just the required set of metadata.

```json
"publications": [
  {
    "metadata": {
      "identifier": "urn:uuid:92ff26c8-55e0-4e86-89fe-62c73ca7f704",
      "removed": {
        "reason": "https://example.com/subscription/removal",
        "detail": "This publication is no longer available in your subscription"
      }
    }
    "images": [
      {
        "href": "https://example.com/cover.jpg",
        "type": "image/jpeg"
      }
    ]
  }
]
```

**Example 2: a specific mean of acquiring the publication is no longer available**

```json
"links": [
  {
    "href": "https://example.com/download.pdf"
    "type": "application/pdf",
    "rel": "http://opds-spec.org/acquisition"
    "properties": {
      "removed": {
        "detail": "PDF is no longer available as a download format for example.com"
      }
    }
  },
  {
    "href": "https://example.com/download.epub"
    "type": "application/epub+zip",
    "rel": "http://opds-spec.org/acquisition"
  }
]
```

**Example 3: a license in an ODL feed is no longer available**

```json
"licenses": [
  {
    "metadata": {
      "identifier": "urn:uuid:f7847120-fc6f-11e3-8158-56847afe9799",
      "format": "application/epub+zip",
      "price": {
        "currency": "USD",
        "value": 7.99
      },
      "created": "2022-04-25T12:25:21+02:00",
      "terms": {
        "checkouts": 30,
        "expires": "2024-04-25T12:25:21+02:00",
        "concurrency": 10,
        "length": 5097600
      },
      "protection": {
        "format": [
          "application/vnd.adobe.adept+xml", 
          "application/vnd.readium.lcp.license.v1.0+json"
        ],
        "devices": 6,
        "copy": false,
        "print": false,
        "tts": false
      },
      "removed": {
        "reason": "https://example.com/reason/refund",
        "detail": "This publication was refunded due to a faulty EPUB that doesn't meet accessibility criteria"
      }, 
    },
    "links": [
      {
        "rel": "http://opds-spec.org/acquisition/borrow",
        "href": "https://example.com/get{?id,checkout_id,expires,patron_id,passphrase,hint,hint_url,notification_url}",
        "type": "application/vnd.readium.license.status.v1.0+json",
        "templated": true
      },
      {
        "rel": "self",
        "href": "https://example.com/status/294024",
        "type": "application/vnd.odl.info+json"
      }
    ]
  }
]
```

**Reply by HadrienGardeur** [reply]
Since naming things is usually the most difficult topic, here are some thoughts and background info :

- in OAI-PMH, there's a `status` with a `deleted` value
- in Atom, there's [RFC 6721](https://datatracker.ietf.org/doc/html/rfc6721) where Atom's `<entry>` is replaced by a `<deleted-entry>` as defined in a new namespace
- that said, I'm not necessarily a big fan of calling this `deleted` because it doesn't necessarily fit our use case
- unlike OAI-PMH or Atom, I don't think that marking the whole publication as deleted would work in our case as we need a much more granular approach to fit our use cases, as illustrated in the examples above
- from a vocabulary standpoint, I also felt that "removed" might be more accurate, for example because a publication might be removed and then added again from a subscription (but English isn't my first language, so I'll let others chime in on that)
- that said, I think we should also explore `unavailable` as an alternative as well, as it might be a better fit to also encompass other use cases: for example a book that is temporarily unavailable because it's currently on preorder
- for `reason` and `detail` I used [RFC 7807](https://datatracker.ietf.org/doc/html/rfc7807) as an influence for the overall design but I felt that `reason` conveyed more meaning than `type`

**Reply by HadrienGardeur** [reply]
I've also collected internal feedback at De Marque (since we use ODL between a number of our products) based on this proposal and this is what people had to say:

- there's a preference for using `unavailable` rather than `removed`
- and in addition to `reason` and `detail`, a number of our devs feel that we should add a timestamp as well

Going back to the first example in this thread, it would look like that:

```json
"unavailable": {
  "reason": "https://example.com/reason/refund",
  "detail": "This publication was refunded due to a faulty EPUB that doesn't meet accessibility criteria",
  "updated": "2023-11-21T11:32:11+02:00"
} 
```

**Reply by HadrienGardeur** [reply]
As a follow-up to our call this week, here's an updated example:

```json
"availability": {
  "state": false,
  "reason": "https://example.com/reason/refund",
  "detail": "This publication was refunded due to a faulty EPUB that doesn't meet accessibility criteria",
  "updated": "2023-11-21T11:32:11+02:00"
} 
```

[Update] Here are some notes from our call:

- @jonathangreen is more comfortable with an `availability` element rather than `removed`/`unavailable` because it's always tricky to handle the absence of a property when harvesting content
- we also discussed about using multiple reasons, but this idea was rejected because it could be confusing with reasons that could contradict one another
- @killerpuppytails is in favour of use `state` to avoid a conflict with `status`
- While looking for inspiration from [RFC 7807](https://datatracker.ietf.org/doc/html/rfc7807) @tdilauro pointed out the difference between `title` and `detail` where `title` would contain a shorter human readable description and `detail` a more extensive one. We agreed that for now, we'll only include `detail` but would consider an equivalent of `title` if it ever became a requirement.
- @Apophenia also pointed out that we might want to support something like `instance` from [RFC 7807](https://datatracker.ietf.org/doc/html/rfc7807) in the future, since it allows to point to a specific explanation/context per publication/license

**Reply by HadrienGardeur** [reply]
Looking at our current schema, there's a potential conflict when using `availability` on a link: https://github.com/opds-community/drafts/blob/master/schema/properties.schema.json#L80-L121

- in this context, availability is meant for the end-user and whether they can borrow/buy a publication immediately
- we also use `state` rather than `status` in this context, with a documented list of values: `available`, `unavailable`, `reserved` and `ready`
- this also contains timestamps (`since` and `until`)

This wouldn't be a conflict when used in:

- publication metadata
- or license metadata

**Reply by HadrienGardeur** [reply]
@barmintor suggested that we could lean into this potential conflict rather than avoid it, which would mean:

- using `state` as it's currently defined in our JSON Schema with an enum rather than a boolean
- and using `since` for what we were considering for `updated`

I think that overall this could be a good approach:

- having a `reason` and a `detail` could be useful for end-users and OPDS clients
- using two timestamps (`since` and `until`) could also prove useful for temporarily unavailable titles (for example when you've reached your maximum budget for the month in a cost per circulation model)

The only part that I have mixed feeling about is the use of an enum of strings/tokens where some values could be unknown/undefined in the context of the availability of a publication/license: `ready` and `reserved`.
I guess that's the price to pay for consistency and we could control this in our JSON Schema with additional checks on `state` within a publication and/or license metadata.

If we go back to our initial examples:

**Example 1: the whole publication has been removed**

```json
"publications": [
  {
    "metadata": {
      "identifier": "urn:uuid:92ff26c8-55e0-4e86-89fe-62c73ca7f704",
      "availability": {
        "state": "unavailable",
        "reason": "https://example.com/subscription/removal",
        "detail": "This publication is no longer available in your subscription",
        "since": "2023-11-21T11:32:11+02:00"
      }
    }
    "images": [
      {
        "href": "https://example.com/cover.jpg",
        "type": "image/jpeg"
      }
    ]
  }
]
```

**Example 2: a specific mean of acquiring the publication is no longer available**

```json
"links": [
  {
    "href": "https://example.com/download.pdf"
    "type": "application/pdf",
    "rel": "http://opds-spec.org/acquisition"
    "properties": {
      "availability": {
        "state": "unavailable",
        "detail": "PDF is no longer available as a download format for example.com",
        "since": "2023-11-21T11:32:11+02:00"
      }
    }
  },
  {
    "href": "https://example.com/download.epub"
    "type": "application/epub+zip",
    "rel": "http://opds-spec.org/acquisition"
  }
]
```

**Example 3: a license in an ODL feed is no longer available**

```json
"licenses": [
  {
    "metadata": {
      "identifier": "urn:uuid:f7847120-fc6f-11e3-8158-56847afe9799",
      "format": "application/epub+zip",
      "price": {
        "currency": "USD",
        "value": 7.99
      },
      "created": "2022-04-25T12:25:21+02:00",
      "terms": {
        "checkouts": 30,
        "expires": "2024-04-25T12:25:21+02:00",
        "concurrency": 10,
        "length": 5097600
      },
      "protection": {
        "format": [
          "application/vnd.adobe.adept+xml", 
          "application/vnd.readium.lcp.license.v1.0+json"
        ],
        "devices": 6,
        "copy": false,
        "print": false,
        "tts": false
      },
      "availability": {
        "state": "unavailable",
        "reason": "https://example.com/reason/refund",
        "detail": "This publication was refunded due to a faulty EPUB that doesn't meet accessibility criteria",
        "since": "2023-11-21T11:32:11+02:00"
      } 
    },
    "links": [
      {
        "rel": "http://opds-spec.org/acquisition/borrow",
        "href": "https://example.com/get{?id,checkout_id,expires,patron_id,passphrase,hint,hint_url,notification_url}",
        "type": "application/vnd.readium.license.status.v1.0+json",
        "templated": true
      },
      {
        "rel": "self",
        "href": "https://example.com/status/294024",
        "type": "application/vnd.odl.info+json"
      }
    ]
  }
]
```

**Comment by barmintor** [comment]
```json
"link" : [
 {
  "rel": "http://opds-spec.org/acquisition/availability",
  "state": "unavailable",
  "reason": {
    "id": "https://example.com/reason/refund",
    "description": "This publication was refunded due to a faulty EPUB that doesn't meet accessibility criteria",
   },
  "since": "2023-11-21T11:32:11+02:00"
} 
]
```

**Comment by HadrienGardeur** [comment]
I've created a draft PR at https://github.com/opds-community/drafts/pull/73

It's still missing quite a few things:

- I'd like to have an enum of well-known reasons, event if we allow any URI for the sake of extensibility
- I haven't added this new JSON Schema to publications or licenses metadata yet
- and we're missing a JSON Schema for ODL anyway so licenses are not covered yet

For our next call, we really need to focus on:

- documenting a first list of well-known reasons (I'll come up with a first proposal)
- make sure that @tdilauro and @jonathangreen are comfortable with the latest proposal

**Comment by HadrienGardeur** [comment]
For the list of reasons, I'd like to use the opds.io domain with a pattern like that: https://registry.opds.io/reason#XYZ

A document with all the proper anchor would be published at that URL and provide a registry of reasons.

Here's an initial list of potential reasons:

| URL | Description |
| ---- | ------------ |
| https://registry.opds.io/reason#exhausted | Indicates that a publications and/or all of the usage allowed by its licenses have been exhausted. |
| https://registry.opds.io/reason#expired | Indicates that a publications and/or all of its licenses have expired. |
| https://registry.opds.io/reason#onHold | Indicates that a publication is temporarily unavailable because it's on hold. |
| https://registry.opds.io/reason#preordered | Indicates that a publication is temporarily unavailable because it hasn't been released yet. |
| https://registry.opds.io/reason#refunded | Indicates that a publication or one of its licenses has been refunded. |
| https://registry.opds.io/reason#removed | Indicates that a publication or one of its licenses has been removed from a subscription or from distribution in general. |

**Reply by HadrienGardeur** [reply]
During our call @tdilauro suggested that for each reason we should also document:

- the relevant `state` values
- the context (publication, link and/or license)
- along with use cases and examples

**Reply by HadrienGardeur** [reply]
I've set up a new repo for all future registries, along with an initial list of values for reasons at https://registry.opds.io/reason

It's still missing context, use cases and examples but I'll add them as well in future updates.

I'll most likely use this opportunity to document other useful registries for the OPDS family of specifications, including (but not limited to) link relationships and properties.

**Comment by jonathangreen** [comment]
@HadrienGardeur I've been thinking about some examples of this for publications where an item has been removed from a subscription. In that case, sometimes we may know about the subscription removal in advance. It would be nice to be able to communicate this as part of the `availability` collection. This could be communicated using a future date for the `since` timestamp. 

For example something like this, assuming the current date is 2023-11-10. This would indicate that the title will be removed from a subscription in 10 days:
```json
{
  "metadata": {
    "@type": "http://schema.org/Book",
    "identifier": "urn:uuid:blah",
    "availability": {
      "state": "unavailable",
      "reason": "https://registry.opds.io/reason#removed",
      "detail": "This publication is no longer available in your subscription",
      "since": "2023-11-20T20:28:13Z"
    }
  }
}
```

In this case though, I don't think that `since` is a very good name for this timestamp, since it implies a time that happened in the past. I'd like to suggest that we rename `since` -> `start` and `until` -> `end`, to make these timestamps use more flexible.

That would make the updated example look like this:
```json
{
  "metadata": {
    "@type": "http://schema.org/Book",
    "identifier": "urn:uuid:blah",
    "availability": {
      "state": "unavailable",
      "reason": "https://registry.opds.io/reason#removed",
      "detail": "This publication is no longer available in your subscription",
      "start": "2023-11-20T20:28:13Z"
    }
  }
}
```

**Reply by HadrienGardeur** [reply]
@jonathangreen the current model is focused entirely on the **present** rather than past or future states. The only info that we currently convey about past/future states is through the presence of `since` and `until` which hint at state changes.

The way we've named things and re-used existing elements was very much a balancing act as well, as we can see from discussions on this page.

With the current proposal, you could indicate that:

- a publication is currently available (`"state": "available"`)
- but that there's an expected change in the future (using `until`)

I don't think that the current element can represent future changes completely, this would require a different element IMO. Listing future states also opens a can of worms: 

- What happens if the future state is incompatible with the present one? 
- How do you deal with multiple future states when they're incompatible with one another?
- How do you deal with canceled state changes?

**Reply by tdilauro** [reply]
@HadrienGardeur:

> With the current proposal, you could indicate that:
> 
> - a publication is currently available ("state": "available")
> - but that there's an expected change in the future (using until)

What if the `since` date/time has not yet arrived?
